linkerd/main/src/main/scala/io/buoyant/linkerd/Main.scala

```diff
@@ -95,17 +97,36 @@ object Main extends App {
     server: Server.Initializer,
     name: Path
   ): Closable = {
+    def addrs: Seq[InetSocketAddress] = {
+      if (server.ip.getHostAddress().equals("0.0.0.0")) {
+        val a = for {
+          interface <- NetworkInterface.getNetworkInterfaces()
+          if interface.isUp()
+        } yield for {
+          inet <- interface.getInetAddresses()
+          if !inet.isLoopbackAddress()
+        } yield new InetSocketAddress(inet.getHostAddress(), server.port)
+        a.flatten.toSeq
+      } else {
+        Seq(server.addr)
+      }
+    }
+
     announcers0.filter { case (pfx, _) => name.startsWith(pfx) } match {
       case Nil =>
         log.warning("no announcer found for %s", name.show)
         Closable.nop
 
       case announcers =>
-        val closers = announcers.map {
+        val closers: Seq[Closable] = announcers.map {
           case (prefix, announcer) =>
-            log.info("announcing %s as %s to %s", server.addr, name.show, announcer.scheme)
-            announcer.announce(server.addr, name.drop(prefix.size))
-        }
+            for {
+              addr <- addrs
+            } yield {
+              log.info("announcing %s as %s to %s", addr, name.show, announcer.scheme)
+              announcer.announce(addr, name.drop(prefix.size))
+            }
+          }.flatten
```

Comment:
you can do `announcers.flatMap { ... }` instead of `announcers.map { ... }.flatten`